consensus/src/main/kotlin/maru/consensus/statetransition/BlockValidator.kt

```diff
@@ -0,0 +1,104 @@
+/*
+   Copyright 2025 Consensys Software Inc.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+package maru.consensus.statetransition
+
+import maru.consensus.SealVerifier
+import maru.consensus.Spec
+import maru.core.BeaconBlock
+import maru.core.BeaconBlockBody
+import maru.core.BeaconState
+import maru.executionlayer.client.ExecutionLayerClient
+import tech.pegasys.teku.infrastructure.async.SafeFuture
+
+interface BlockValidator {
+  sealed interface ValidationResult {
+    data class Valid(
+      val postBeaconState: BeaconState,
+    ) : ValidationResult
+
+    data class Invalid(
+      val reason: String,
+    ) : ValidationResult
+  }
+
+  fun validateBlock(
+    block: BeaconBlock,
+    preState: BeaconState,
+  ): SafeFuture<ValidationResult>
+}
+
+class BlockValidatorImpl(
+  private val spec: Spec,
+  private val sealVerifier: SealVerifier,
+  private val executionLayerClient: ExecutionLayerClient,
+) : BlockValidator {
+  override fun validateBlock(
+    block: BeaconBlock,
+    preState: BeaconState,
+  ): SafeFuture<BlockValidator.ValidationResult> {
+    if (block.beaconBlockHeader.number != preState.latestBeaconBlockHeader.number + 1u) {
+      return SafeFuture.completedFuture(
+        BlockValidator.ValidationResult.Invalid("Block number is not the next block number"),
+      )
+    }
+
+    if (!block.beaconBlockHeader.parentRoot.contentEquals(preState.latestBeaconBlockRoot)) {
+      return SafeFuture.completedFuture(BlockValidator.ValidationResult.Invalid("Parent root does not match"))
+    }
+
+    val beaconBlockBodyRoot = computeBeaconBlockBodyHash(block.beaconBlockBody)
+
+    val postBeaconStateForRootCalculation =
+      preState.copy(
+        latestBeaconBlockHeader = block.beaconBlockHeader.copy(stateRoot = ByteArray(0)),
+        latestBeaconBlockRoot = beaconBlockBodyRoot,
+      )
+    val stateRoot = computeBeaconStateHash(postBeaconStateForRootCalculation)
+
+    if (!block.beaconBlockHeader.stateRoot.contentEquals(stateRoot)) {
+      return SafeFuture.completedFuture(BlockValidator.ValidationResult.Invalid("State root does not match"))
+    }
+
+    val postBeaconState =
+      preState.copy(
+        latestBeaconBlockHeader = block.beaconBlockHeader,
+        latestBeaconBlockRoot = beaconBlockBodyRoot,
+      )
```

Comment:
The creation of the post beacon state should come out into its own state transition class as the same logic is needed during block creation